src/lib.rs

```diff
@@ -98,23 +111,76 @@ where
 #[instrument]
 async fn serve_ohttp_relay(
     req: Request<Incoming>,
-    gateway_origin: &GatewayUri,
+    default_gateway: &GatewayUri,
+    prober: &Prober,
 ) -> Result<Response<BoxBody<Bytes, hyper::Error>>, hyper::Error> {
-    let path = req.uri().path();
-    let mut res = match (req.method(), path) {
+    let mut res = match (req.method(), req.uri().path()) {
         (&Method::OPTIONS, _) => Ok(handle_preflight()),
         (&Method::GET, "/health") => Ok(health_check().await),
-        (&Method::POST, "/") => handle_ohttp_relay(req, gateway_origin).await,
+        (&Method::POST, _) => match parse_gateway_uri(&req, default_gateway, prober).await {
+            Ok(gateway_uri) => handle_ohttp_relay(req, &gateway_uri).await,
+            Err(e) => Err(e),
+        },
         #[cfg(any(feature = "connect-bootstrap", feature = "ws-bootstrap"))]
-        (&Method::CONNECT, _) | (&Method::GET, _) =>
-            crate::bootstrap::handle_ohttp_keys(req, gateway_origin).await,
-        _ => Err(Error::NotFound),
+        (&Method::GET, _) | (&Method::CONNECT, _) =>
+            match parse_gateway_uri(&req, default_gateway, prober).await {
+                Ok(gateway_uri) => crate::bootstrap::handle_ohttp_keys(req, &gateway_uri).await,
+                Err(e) => Err(e),
+            },
+        _ => Err(Error::MethodNotAllowed),
```

Comment:
We're not producing these responses according to spec because they're not producing Allow headers afaict.

"The server must generate an [Allow](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Allow) header in a 405 response with a list of methods that the target resource currently supports." [src](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405).

I do think for the function of our server MethodNotAllowed could make sense, but don't have a strong preference over NotFound here.

Reply:
If we were to do that I think we'd want to subdivide method handlers and then path handlers. A MethodNotAllowed response should only really return on a method check, otherwise a change could introduce a tuple that created some new method that might be allowed but only on a specific path, and MethodNotAllowed would still be returned if only the path were not hit.